File: Dsa-day356.py
class Solution:
    def uniqueXorTriplets(self, nums: List[int]) -> int:
        max_val = max(nums)
        bits = 1
        while (1 << bits) <= max_val:
            bits += 1
        limit = 1 << bits

        unique_vals = sorted(set(nums))

        pair_xor = [False] * limit
        u_count = len(unique_vals)
        for i in range(u_count):
            for j in range(i, u_count):
                pair_xor[unique_vals[i] ^ unique_vals[j]] = True

        triple_xor = [False] * limit
        for p in range(limit):
            if pair_xor[p]:
                for v in unique_vals:
                    triple_xor[p ^ v] = True

        return sum(triple_xor)

File: Dsa-day356.cpp
class Solution {
public:
    int uniqueXorTriplets(vector<int>& nums) {
        int maxVal = 0;
        for (int x : nums) {
            maxVal = max(maxVal, x);
        }
        int bits = 1;
        while ((1 << bits) <= maxVal) {
            bits++;
        }
        int limit = 1 << bits;

        vector<bool> present(limit, false);
        for (int x : nums) {
            present[x] = true;
        }

        vector<int> uniqueVals;
        for (int v = 0; v < limit; v++) {
            if (present[v]) {
                uniqueVals.push_back(v);
            }
        }

        vector<bool> pairXor(limit, false);
        int uCount = uniqueVals.size();
        for (int i = 0; i < uCount; i++) {
            for (int j = i; j < uCount; j++) {
                pairXor[uniqueVals[i] ^ uniqueVals[j]] = true;
            }
        }

        vector<bool> tripleXor(limit, false);
        for (int p = 0; p < limit; p++) {
            if (pairXor[p]) {
                for (int i = 0; i < uCount; i++) {
                    tripleXor[p ^ uniqueVals[i]] = true;
                }
            }
        }

        int count = 0;
        for (int v = 0; v < limit; v++) {
            if (tripleXor[v]) {
                count++;
            }
        }
        return count;
    }
};